Improve generating Custom entry function

Ayush1325 opened the pull request:
This commit is aimed at making compiler-generated entry functions (Basically just C `main` right now) more generic so other targets can do similar things for custom entry. This was initially implemented as part of https://github.com/rust-lang/rust/pull/100316.

Currently, this moves the entry function name and Call convention to the target spec.

Signed-off-by: Ayush Singh <ayushsingh1325@gmail.com>

Ayush1325 committed:
Improve generating Custom entry function

This commit is aimed at making compiler generated entry functions
(Basically just C `main` right now) more generic so other targets can do
similar things for custom entry. This was initially implemented as part
of https://github.com/rust-lang/rust/pull/100316.

Currently, this moves the entry function name and Call convention to the
target spec.

Signed-off-by: Ayush Singh <ayushsingh1325@gmail.com>

wesleywiser commented on compiler/rustc_codegen_llvm/src/declare.rs:
```diff
@@ -90,6 +90,28 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
+        unnamed: llvm::UnnamedAddr,
+        fn_type: &'ll Type,
+    ) -> &'ll Value {
+        let visibility = if self.tcx.sess.target.default_hidden_visibility {
+            llvm::Visibility::Hidden
```
Does it make sense for the entry function to ever be hidden? It also seems a bit weird for the `default_hidden_visiblility` field to control this since (as far as I know) there's no way for the user to override this.

Ayush1325 replied:
Well, I have kept it there since this function is a copy of `declare_cfn` (which was originally used to create the entry function).

Ayush1325 replied:
@wesleywiser After reading a bit on [visibility-styles](https://llvm.org/docs/LangRef.html#visibility-styles), I think we can safely always used `llvm::Visibility::Default` for entry function. Should I make the change?

wesleywiser replied:
I think what you have is fine and since `declare_cfn` was doing the same thing, I don't see a strong reason to change it at this time.

wesleywiser commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1668,6 +1669,12 @@ pub struct TargetOptions {
     /// since this is most common among tier 1 and tier 2 targets.
     pub supports_stack_protector: bool,
+
+    // The name of entry function.
+    pub entry_name: StaticCow<str>,
+
+    // The ABI of entry function.
```
Nit: I think it would be useful for these comments to mention the default values.

Ayush1325 commented:
> And now there will be 3 copy&pasted matches instead of one? Maybe there is a other way.

Can you elaborate on what you mean by "copy&pasted matches"?

klensy commented:
> Can you elaborate on what you mean by "copy&pasted matches"?

That big match from `llvm_cconv` and it's from\into string conversions. Maybe write separate FromStr impl for Conv instead of placing that code into rustc_target.

wesleywiser commented:
Thanks @Ayush1325, this looks good to me!

@bors r+

bors commented:
:pushpin: Commit 9f0a8620bd7d325e6d42417b08daff3e55cb88f6 has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser 9f0a8620bd7d325e6d42417b08daff3e55cb88f6 -->
<!-- homu: {"type":"Approved","sha":"9f0a8620bd7d325e6d42417b08daff3e55cb88f6","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented on compiler/rustc_codegen_llvm/src/context.rs:
```diff
@@ -570,8 +570,14 @@ impl<'ll, 'tcx> MiscMethods<'tcx> for CodegenCx<'ll, 'tcx> {
+        if self.get_declared_value(entry_name).is_none() {
+            Some(self.declare_entry_fn(
+                entry_name,
+                self.sess().target.entry_abi.into(),
```
rustc_codegen_gcc and rustc_codegen_cranelift also need this change.

Ayush1325 replied:
Should I add that in a different PR or update this one?

bjorn3 replied:
A separate PR is fine by me. But if you want to update this PR, I can review it too.

Ayush1325 replied:
Ok, So I'm not sure how to approach gcc. As is visible [here](https://github.com/rust-lang/rust/blob/357f66072959f494d5fa2e6a269c6100eed703c1/compiler/rustc_codegen_gcc/src/declare.rs#L68), the entry function signature and call convention is hard-coded there. So there is not much point in changing the entry name until the `TODOs` are fixed.

bjorn3 replied:
For cg_clif the change would be in src/main_shim.rs. As for cg_gcc cc @antoyo.

antoyo replied:
I believe the same change as you did for LLVM applies: it's just that the function type is hard-coded for now. You can keep it that way.

Ayush1325 committed:
Use custom entry name in gcc

This is a continuation of 9f0a8620bd7d325e6d42417b08daff3e55cb88f6 for
gcc.

Signed-off-by: Ayush Singh <ayushsingh1325@gmail.com>

Ayush1325 committed:
Use custom entry name in cranelift

This is a continuation of 9f0a8620bd7d325e6d42417b08daff3e55cb88f6 for
cranelift.

Signed-off-by: Ayush Singh <ayushsingh1325@gmail.com>

Ayush1325 commented:
@bjorn3 I have tried doing the same stuff for cranelift. However, I'm not sure how to map the call convention with `entry_abi` since the CallConv in cranelift are unfamiliar to me.

bjorn3 commented:
You could extract the match at https://github.com/rust-lang/rust/blob/9340e5c1b9dee53fd32a18f7bfb54faabfe00b7b/compiler/rustc_codegen_cranelift/src/abi/mod.rs#L25 into a function and then use this function.

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 2436dff7727732a9f55b1c80888be66ada8e288f has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 2436dff7727732a9f55b1c80888be66ada8e288f -->
<!-- homu: {"type":"Approved","sha":"2436dff7727732a9f55b1c80888be66ada8e288f","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/104361#issuecomment-1317416027) for pull requests below priority 1. This pull request will be tested once the tree is reopened.

Ayush1325 commented:
@antoyo Can you check out 6dfe2395c9f26076e339ea40c269f4b23f8c22c8 as well?

bjorn3 commented:
I think it is fine.

antoyo commented:
> @antoyo Can you check out [6dfe239](https://github.com/rust-lang/rust/commit/6dfe2395c9f26076e339ea40c269f4b23f8c22c8) as well?

That looks fine to me!
